tsl/src/continuous_aggs/refresh.c

```diff
@@ -302,19 +287,39 @@ continuous_agg_refresh(PG_FUNCTION_ARGS)
 		refresh_window.end = ts_time_value_from_arg(PG_GETARG_DATUM(2),
 													get_fn_expr_argtype(fcinfo->flinfo, 2),
 													refresh_window.type);
+	continuous_agg_refresh_internal(cagg, &refresh_window);
+	PG_RETURN_VOID();
+}
 
-	if (refresh_window.start >= refresh_window.end)
+void
+continuous_agg_refresh_internal(ContinuousAgg *cagg, InternalTimeRange *refresh_window)
+{
+	Catalog *catalog = ts_catalog_get();
+	Hypertable *cagg_ht;
+	int32 mat_id = cagg->data.mat_hypertable_id;
+	PreventCommandIfReadOnly(REFRESH_FUNCTION_NAME);
+
+	/* Prevent running refresh if we're in a transaction block since a refresh
+	 * can run two transactions and might take a long time to release locks if
+	 * there's a lot to materialize. Strictly, it is optional to prohibit
+	 * transaction blocks since there will be only one transaction if the
+	 * invalidation threshold needs no update. However, materialization might
+	 * still take a long time and it is probably best for conistency to always
+	 * prevent transaction blocks.  */
+	PreventInTransactionBlock(true, REFRESH_FUNCTION_NAME);
+
+	if (refresh_window->start >= refresh_window->end)
 		ereport(ERROR,
 				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
 				 errmsg("invalid refresh window"),
 				 errhint("The start of the window must be before the end.")));
 
-	refresh_window = compute_bucketed_refresh_window(&refresh_window, cagg->data.bucket_width);
+	*refresh_window = compute_bucketed_refresh_window(refresh_window, cagg->data.bucket_width);
```

Comment:
I suggest this be a new variable as opposed to changing the passed in pointed to one.

Reply:
fixed.